quickwit-search/src/search_client_pool.rs

```diff
@@ -241,33 +247,44 @@ impl SearchClientPool {
             }
         }
 
-        // Sort job
-        jobs.sort_by(|left, right| {
+        // sort jobs by cost desc, split_id asc.
+        //
+        // By keeping split_id in comparision, we minimize cost for next rendez_vous sort when
+        // two consecutive split_id(s) are the same.
+        jobs.sort_unstable_by(|left, right| {
             // sort_by_key does not work here unfortunately
             job_order_key(left).cmp(&job_order_key(right))
         });
 
+        // Compute load (per node) threshold
+        let total_cost: Cost = jobs.iter().map(|v| v.cost()).sum();
+        let load_per_node_threshold =
+            ((total_cost as f64 / nodes.len() as f64) * (1.0 + LOAD_MARGIN)).ceil() as u64;
+        let load_per_node_threshold = load_per_node_threshold.max(MIN_LOAD_THRESHOLD);
```

Comment:
So we have an edge case here if we have a lot of nodes and 1 large split.

The `MIN_LOAD_THRESHOLD` seems to be helping here but it does not seem very good at that.
I suggest we change the MIN_LOAD_THRESHOLD to the largest cost within the splits.

Reply:
I wonder if we can have a more solid math background here.

For instance if we define,

`load_per_node_threshold = ceil([total_cost / nodes] + MAX_NODE_COST)`

and we never add a split to a node if it would result to a load exceeding load_per_node_threshold, I think we have some interesting property:

Without extra checks, the algorithm would be able to assign all split to nodes.
Proof:

Let's say that we are left with a split that is unassigned.
It means that all nodes are loaded with at  `load > load_per_node_threshold - missing_node`.
(Otherwise we could allocate the missing node).

In other word, the overall load in the nodes would be greater than 
```
    (load_per_node_threshold - missing_node_load) * num_nodes
> total_cost + num_nodes * (max_nodes  - missing_node_load)
> total_cost
```
which is impossible...

We can even probably relax that condition, since we sorted the node per cost.
For instance, I suggest the following formula

```
base_load_per_node_threshold = (total_cost as f64) / nodes.len() as f64;
optimality_margin = base_load_per_node_threshold * 0.2
feasability_margin = splits[min(num_nodes, splits.len - 1)].cost # 

load_per_node_threshold
    =  max(
             max_node_cost
             base_load_per_node_threshold + max(optimality_margin, feasability_margin)
    )
     
```

Reply:
@fmassot @linxGnu  what do you think?

Reply:
@fulmicoton I think your suggestion is good in case of having several split with "super" cost. This would make imbalance cost between node.

There is edge case for the formula above:
```
costs: 5000, 2500, 100, 100, 100, 100, 100, 100
total_cost: 8100
num_nodes: 4
base: 8100 / 4 = 2025

-> load_per_node_threshold
    =  max(
             max_node_cost
             base_load_per_node_threshold + max(optimality_margin, feasability_margin)
    )
    = max(5000, 2025 + max(2025 * 0.2, 100)) = 5000

Distribution:
node[0].cost = 5000
node[1].cost = 2500 + 100 + 100 = 2700
node[2].cost = 100 + 100 + 100 + 100 = 400
```

While original algorithm by fmassot (looks better in my point of view):
```
load_per_node = 2025 + 2025 * 0.2 = 2430

node[0].cost = 5000
node[1].cost = 2500
node[2].cost = 600
```

Actually, all are heuristic and have edge cases imho. How about modify your formula a bit:
```
load_per_node_threshold
    =  base_load_per_node_threshold + max(optimality_margin, feasability_margin)
```
This one removes the impact of super cost (max_node_cost, like above, it's 5000).